pkg/typesystem/typesystem_test.go

```diff
@@ -12,6 +12,555 @@ import (
 	"github.com/openfga/openfga/pkg/testutils"
 )
 
+type relationDetails struct {
+	hasEntrypoints bool
+	hasLoop        bool
+}
+
+func TestHasEntrypoints(t *testing.T) {
+	tests := map[string]struct {
+		model         string
+		inputType     string
+		inputRelation string
+		expectError   string
+		expectDetails *relationDetails
+	}{
+		`undefined_input_type`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [folder]`,
+			inputType:     "unknown",
+			inputRelation: "viewer",
+			expectError:   "undefined type definition for 'unknown#viewer'",
+		},
+		`undefined_input_relation`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [folder]`,
+			inputType:     "document",
+			inputRelation: "unknown",
+			expectError:   "undefined type definition for 'document#unknown'",
+		},
+		`undefined_type_in_assignable_type`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [unknown#editor]`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectError:   "undefined type definition for 'unknown#editor'",
+		},
+		`undefined_relation_in_assignable_type`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: [document#unknown]`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectError:   "undefined type definition for 'document#unknown'",
+		},
+		`undefined_computed_userset`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: unknown`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectError:   "undefined type definition for 'document#unknown'",
+		},
+		`undefined_tupleset`: {
+			model: `
+			model
+				schema 1.1
+			type document
+				relations
+					define viewer: viewer from unknown`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectError:   "undefined type definition for 'document#unknown'",
+		},
+		`undefined_computed_relation_on_tupleset_target`: {
+			model: `
+			model
+				schema 1.1
+			type folder
+				relations
+					define owner: [user]
+			type document
+				relations
+					define parent: [folder]
+					define viewer: viewer from parent`,
+			inputType:     "document",
+			inputRelation: "viewer",
+			expectDetails: &relationDetails{false, false},
```

Comment:
This is an invalid model - no entrypoints shouldn't be applicable.

Reply:
added a TODO to `undefined_computed_relation_on_tupleset_target`